numpy/core/src/multiarray/ctors.c

```diff
@@ -1072,7 +1072,17 @@ PyArray_NewFromDescr_int(PyTypeObject *subtype, PyArray_Descr *descr, int nd,
             data = npy_alloc_cache(nbytes);
         }
         if (data == NULL) {
-            PyErr_NoMemory();
+            PyObject *errmsg = PyUString_FromString("Array shape:(");
+            for (i = 0; i < fa->nd-1; i++) {
+                PyUString_ConcatAndDel(&errmsg,
+                    PyUString_FromFormat("%ld,",fa->dimensions[i]));
+            }
+            PyUString_ConcatAndDel(&errmsg,
+                PyUString_FromFormat("%ld); Data type:",fa->dimensions[i]));
+            PyUString_ConcatAndDel(&errmsg,
+                PyUString_FromFormat("%s.",fa->descr->typeobj->tp_name));
+            PyErr_SetObject(PyExc_MemoryError, errmsg);
```

Comment:
Unforunately you need to check every single function call here for a `NULL` return value, else you risk producing a segfault.

This is why I've been increasingly leaning towards doing the string formatting in python.